sources/osgShadow/shaders/shadowLinearSoft.glsl

```diff
@@ -1,23 +1,45 @@
+#pragma include "floatFromTex.glsl"
+
+
+#ifdef _OUT_DISTANCE
+#define OPT_ARG_outDistance ,out float outDistance
+#else
+#define OPT_ARG_outDistance
+#endif
 
 // simulation of texture2Dshadow glsl call on HW
 // http://codeflow.org/entries/2013/feb/15/soft-shadow-mapping/
 float texture2DCompare(const in sampler2D depths, const in vec2 uv, const in float compare, const in vec4 clampDimension){
-
     float depth = getSingleFloatFromTex(depths, clamp(uv, clampDimension.xy, clampDimension.zw));
-    return step(compare, depth);
-
+    return compare - depth;
 }
 
 // simulates linear fetch like texture2d shadow
-float texture2DShadowLerp(const in sampler2D depths, const in vec4 size, const in vec2 uv, const in float compare, const in vec4 clampDimension){
+float texture2DShadowLerp(
+    const in sampler2D depths,
+    const in vec4 size,
+    const in vec2 uv,
+    const in float compare,
+    const in vec4 clampDimension
+    OPT_ARG_outDistance){
+
+    vec2 f = fract(uv * size.xy + 0.5);
+    vec2 centroidUV = floor(uv * size.xy + 0.5) * size.zw;
+
+    float lb = texture2DCompare(depths, centroidUV + size.zw * vec2(0.0, 0.0), compare, clampDimension);
+    float lt = texture2DCompare(depths, centroidUV + size.zw * vec2(0.0, 1.0), compare, clampDimension);
+    float rb = texture2DCompare(depths, centroidUV + size.zw * vec2(1.0, 0.0), compare, clampDimension);
+    float rt = texture2DCompare(depths, centroidUV + size.zw * vec2(1.0, 1.0), compare, clampDimension);
+
+#ifdef _OUT_DISTANCE
+    outDistance = mix(mix(lb, lt, f.y), mix(rb, rt, f.y), f.x);
+#endif
 
-    vec2 f = fract(uv*size.xy+0.5);
-    vec2 centroidUV = floor(uv*size.xy+0.5)*size.zw;
+    lb = step(lb, 0.0);
```

Comment:
can't tell if it makes a difference or not, I did the step vec4 but not the early out (dunno if it's worth it)